### PR TITLE
feat: get file author from github

### PR DESCRIPTION
**Pull request by niktverd**

### CODE REVIEW

**Comment by niktverd** [comment]
manually added
---
author: {'avatar':'https://avatars.githubusercontent.com/u/43607074?v=4','email':'alehaymkm@gmail.com','login':'alehay','name':'Aleksei Kudakov','url':'https://github.com/alehay'}
__system: {"dislikeVariants":["Нет ответа на мой вопрос","Рекомендации не помогли","Содержание не соответствует заголовку","Другое"]}
contributors: [{'avatar':'https://avatars.githubusercontent.com/u/43607074?v=4','email':'alehaymkm@gmail.com','login':'alehay','name':'Aleksei Kudakov','url':'https://github.com/alehay','date':'2021-11-16T09:31:43Z'}]
---

automatically added
---
__system: {"dislikeVariants":["Нет ответа на мой вопрос","Рекомендации не помогли","Содержание не соответствует заголовку","Другое"]}

author: {'avatar':'https://avatars.githubusercontent.com/u/56478986?v=4','email':'56478986+gayratvg@users.noreply.github.com','login':'gayratvg','name':'Gayrat Vlasov_2','url':'https://github.com/gayratvg','date':'2022-02-21T09:55:38Z'}
contributors: [{'avatar':'https://avatars.githubusercontent.com/u/56478986?v=4','email':'56478986+gayratvg@users.noreply.github.com','login':'gayratvg','name':'Gayrat Vlasov_2','url':'https://github.com/gayratvg','date':'2022-02-21T09:55:38Z'}]
---

File: src/vcs-connector/github.ts [hunk 1]
```diff
@@ -140,12 +141,14 @@ async function getContributorDataByHashCommit(httpClientByToken: Octokit, hashCo
 
     const {avatar_url: avatar, html_url: url, login} = author;
 
+    console.log({author: commit.author});
```
Comment (moki) [inline src/vcs-connector/github.ts:144]: unnecessary logging

File: src/services/utils.ts [hunk 2]
```diff
@@ -149,3 +151,34 @@ export function liquidField(input: string, vars: Record<string, unknown>, path:
...
+export function getContributorArrayFromMetadata(metadata: string[]) {
+    const contributors: Contributor[] = [];
+
+    for (const meta of metadata) {
+        const matchContributors = meta.match(REGEXP_CONTRIBUTORS);
+        try {
+            if (matchContributors) {
+                contributors.push(...parseDataFromJSONString(matchContributors[0]));
+            }
+        } catch (_) {}
```
Comment (moki) [inline src/services/utils.ts:180]: proper error handling is missing

File: src/vcs-connector/github.ts [hunk 3]
```diff
@@ -193,13 +196,28 @@ function addContributorForPath(paths: string[], newContributor: Contributors, ha
...
         const oldContributors = contributorsByPath.get(normalizePath);
 
-        contributorsByPath.set(normalizePath, {
-            contributors: {
-                ...oldContributors?.contributors,
-                ...newContributor,
-            },
-            hasIncludes,
-        });
+        const newContributorKey = Object.keys(newContributor)[0];
```
Comment (moki) [inline src/vcs-connector/github.ts:199]: is it impossible for a `newContributor` object to be empty?

**Review by moki — CHANGES_REQUESTED** [review]

File: src/services/authors.ts [hunk 4]
```diff
@@ -1,24 +1,42 @@
...
+    }
+
+    const contributors = getContributorArrayFromMetadata(newMetadatas);
+
+    if (contributors.length && !matchAuthor) {
+        contributors.sort((a, b) => isEarlier(a.date, b.date) ? -1 : 1);
+        const user = await getAuthorDetails(vcsConnector, contributors[0]);
+
+        if (user) {
+            return `${fileMetadata}${os.EOL}author: ${user}`;
```
Comment (moki) [inline src/services/authors.ts:35]: you don't want to use os.EOL (os specific symbols) if your code will run on one system, and then write data for later to be used somewhere in the other system with different os.

use os agnostic new line symbol

File: src/vcs-connector/github.ts (hunk 1 above)
Comment (niktverd) [reply src/vcs-connector/github.ts:144]: exactly

File: src/services/utils.ts (hunk 2 above)
Comment (niktverd) [reply src/services/utils.ts:180]: sure

File: src/vcs-connector/github.ts (hunk 3 above)
Comment (niktverd) [reply src/vcs-connector/github.ts:199]: Looks likes it cannot be empty. But let's add an additional check

File: src/services/authors.ts (hunk 4 above)
Comment (niktverd) [reply src/services/authors.ts:35]: We already had a talk in a private chat. Shortly, running on `Windows` already inserts `os specific symbol`. Adding os.EOL helps to maintain sameness

File: src/services/utils.ts (hunk 2 above)
Comment (moki) [reply src/services/utils.ts]: please use [logger module](https://github.com/yandex-cloud/yfm-docs/blob/master/src/utils/logger.ts)

warn for `non fatal` errors and `error` for fatal ones.

File: src/services/utils.ts [hunk 5]
```diff
@@ -149,3 +151,36 @@ export function liquidField(input: string, vars: Record<string, unknown>, path:
...
+    const contributors: Contributor[] = [];
+
+    for (const meta of metadata) {
+        const matchContributors = meta.match(REGEXP_CONTRIBUTORS);
+        try {
+            if (matchContributors) {
+                contributors.push(...parseDataFromJSONString(matchContributors[0]));
+            }
+        } catch (err) {
+            console.error(err);
```
Comment (moki) [inline src/services/utils.ts:181]: please use [logger module](https://github.com/yandex-cloud/yfm-docs/blob/master/src/utils/logger.ts)

warn for `non fatal` errors and `error` for fatal ones.

File: src/services/utils.ts [hunk 6]
```diff
@@ -149,3 +151,36 @@ export function liquidField(input: string, vars: Record<string, unknown>, path:
...
+    }
+
+    return new Date(oldDate) > new Date(newDate);
+}
+
+export function parseDataFromJSONString(dataString: string) {
+    return JSON.parse(replaceSingleToDoubleQuotes(dataString));
+}
+
+export function getContributorArrayFromMetadata(metadata: string[]) {
```
Comment (moki) [inline src/services/utils.ts:171]: also thinking about it – i don't think this is utility function, looks more like authors service buisness logic.

Comment (niktverd) [reply src/services/utils.ts]: moved

File: src/services/utils.ts (hunk 5 above)
Comment (niktverd) [reply src/services/utils.ts]: ok

**Review by moki — APPROVED** [review]

**Pull request merged**